### PR TITLE
14 use GitHub runner to automatically pull the latest image from dockerhub to daniels server

### LINKED ISSUE #14: Use Github Runner to automatically pull the latest image from Dockerhub to Daniel's server
Depends on #10

### PR DESCRIPTION
**Pull request by andybaumgar**

### CODE REVIEW

File: env_json_to_dotenv.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# Description: Takes 1 arguement with name of environment variable which contains JSON representation of .env secrets and writes actual .env file.  Allows many GitHub Secrets to be passed through into Docker environment together without hardcoding.
+
+import json
+import sys
+import os
+
+args = sys.argv
+secrets_var = args[1]
```
Comment (Andrew-Dickinson) [inline env_json_to_dotenv.py]: Nit: I usually see this written as `secrets_var = sys.args[1]`

**Comment by andybaumgar** [comment]
I think I'll refactor this to use an entry point script instead of the docker-compose and expand the .env inside the container vs in the GitHub runner using an env variable.  The combo of volumes + weird Dockerfile is kind of confusing.

**Comment by andybaumgar** [comment]
On second thought, to avoid complicating this non-ideal GitHub secrets setup further, I'll just leave it as is.  The alternatives don't seem much better.  Ideally we would just use 3rd party key store to manage our secrets so that the local and server code would work the same, but that seems really overkill for this project.

Appendix: here are the non ideal alternatives to our current solution (.env Docker volume).

1.  Pass in all secrets as JSON encoded environment variable instead of expanding to .env before running the container.  For the server this change could just require refactoring a bit, but for running locally I'd have to write more code to convert the .env to JSON.  Seems very Rube Goldberg.
2. The "dumb" way.  Pass in 16 named command-line ENV arguments to Docker.  We call the app multiple times (testing, deploying, locally) so this would get very verbose.  We'd probably want to wrap the Docker call in a script, but that seems just as complex as option 1.  Also not clear it will work with multi line keys.
3. Keep GitHub Secrets, don't use Docker.  We'd still need a script to remap the GitHub Secrets since they only exist inside a special GitHub context variable.  It would seem weird to hard code our app to fit GitHub Secrets as well.
4. Don't use Docker or GitHub Secrets.  Manually place .env on server where code will run.  Seems messy and non standard.  We'd have to install a bunch of Python stuff on Daniels server and then reinstall if we ever move it.
5. Keep Docker, don't use GitHub secrets.  We'd still have to use volumes to get the .env into Docker.
6. Keep Docker, build locally on server with local .env included in Dockerfile.  Non standard but easier.  Need to SSH into server to update .env.
7. Let Docker read in .env file.  It breaks with multi line keys in a way that's not easily fixable.  The issue page says to just use a key store.

**Comment by andybaumgar** [comment]
I'm remembering Docker actually reads the .env file by default, but the Python process we start wasn't seeing them.  If I could fix this it might be a cleaner option.

**Comment by andybaumgar** [comment]
Ugh, the newlines in the Google key that were giving me problems before also break the Docker .env integration.  One post recommended a key store - basically the ideal option I mentioned above.

**Comment by andybaumgar** [comment]
I removed the docker-compose.yml since it's usage was a bit confusing.  The two docker commands are just a bit longer now.

**Comment by andybaumgar** [comment]
Note: currently the deploy container will exit since we don't pass in credentials.json.  I think we plan to move this to the .env eventually, so I've left it out for now.

File: Dockerfile [hunk 2]
```diff
@@ -4,10 +4,12 @@ FROM python:3.9-slim-buster
 
 WORKDIR /app
 
-# Note: Eventaully many files will be omitted. Images build with current Dockerfile should not be pushed to a public repository.
+COPY requirements.txt requirements.txt
+RUN pip3 install -r requirements.txt
```
Comment (Andrew-Dickinson) [inline Dockerfile:8]: Why not move this command after the `COPY . .` on line 11 below, and save the explicit COPY of `requriments.txt` above on line 7

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,60 @@
+name: CI/CD
```
Comment (Andrew-Dickinson) [inline .github/workflows/ci.yml:1]: Maybe I'm misunderstanding this, but I thought the idea was to have the CI build in the cloud? Is that called "self-hosted" in this file? Or are we trying to do the CI build on Daniel's server?

**Review by Andrew-Dickinson — APPROVED** [review]

File: Dockerfile (hunk 2 above)
Comment (andybaumgar) [reply Dockerfile:8]: @Andrew-Dickinson  it builds faster this way since the pip step is cached when the file is unchanged.

Comment (andybaumgar) [reply Dockerfile:8]: Mostly when building locally.

File: .github/workflows/ci.yml (hunk 3 above)
Comment (andybaumgar) [reply .github/workflows/ci.yml:1]: @Andrew-Dickinson  We use the self hosted runner to execute code on Daniel's server, mostly the Deploy job (pulling the current app image and running it with Docker).  I can run the Test and Build jobs on GitHub hosted runners if you want.  For our case I don't think it really matters, but I could see that being safer if the service was really important.

File: Dockerfile (hunk 2 above)
Comment (Andrew-Dickinson) [reply Dockerfile:8]: Oh right

File: .github/workflows/ci.yml (hunk 3 above)
Comment (andybaumgar) [reply .github/workflows/ci.yml:1]: @Andrew-Dickinson I changed the Test and Build jobs to use GitHub hosted.  I [tested it](https://github.com/andybaumgar/nycmesh-support-bot/actions/runs/2985790994) and every seems to work.  The Test and Build steps are a bit slower, but if we actually care about speeding them up there are probably lots of other Docker optimizations.

**Review by Andrew-Dickinson — APPROVED** [review]

File: env_json_to_dotenv.py (hunk 1 above)
Comment (andybaumgar) [reply env_json_to_dotenv.py]: Ah good point.  Updated it.

Comment (Andrew-Dickinson) [reply env_json_to_dotenv.py]: I'm wrong, it's sys.argv[1]

Comment (andybaumgar) [reply env_json_to_dotenv.py]: Thanks, fixed it!

**Pull request merged**